AX: Move AccessibilityRole into its own header to fix a non-unified build issue in AXTableHelpers.cpp

twilco opened the pull request:
#### e5af8a791832ba5077414e5ac00bbf0a435169cc
<pre>
AX: Move AccessibilityRole into its own header to fix a non-unified build issue in AXTableHelpers.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=297859">https://bugs.webkit.org/show_bug.cgi?id=297859</a>
<a href="https://rdar.apple.com/159107755">rdar://159107755</a>

Reviewed by Joshua Hoffman.

This commit allows use of AccessibilityRole without importing the entirety of AXCoreObject.h.

Renames accessibilityRoleToString to roleToString.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::accessibilityRoleToString): Deleted.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXTableHelpers.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::debugDescription const):
* Source/WebCore/accessibility/AXUtilities.cpp:
(WebCore::roleToString):
* Source/WebCore/accessibility/AXUtilities.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::debugDescriptionInternal const):
* Source/WebCore/accessibility/AccessibilityRole.h: Added.
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper _accessibilityWebRoleAsString]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::debugDescriptionInternal const):

Canonical link: <a href="https://commits.webkit.org/299307@main">https://commits.webkit.org/299307@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de5b171bc493831b22f44cb7a9b542a3da0f0d1b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118568 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28900 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124744 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70627 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c385b1ed-1010-43f0-8fe4-b89d7faa571b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120446 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38945 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46831 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89991 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59544 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a6018b2c-b67a-4dd9-a69e-827666a724eb) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121521 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31019 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106305 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70495 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2d54d346-2653-4b45-9537-212283c443f9) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30074 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24416 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68400 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100457 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24607 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127806 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45475 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34305 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98623 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45839 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98407 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25017 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43846 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21843 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45345 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51023 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44808 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48155 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46495 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

minorninth reviewed (approved):
Thanks for doing this!

fleizach commented on Source/WebCore/accessibility/AccessibilityRole.h:
```diff
@@ -0,0 +1,445 @@
+
+using AccessibilityRoleSet = HashSet<AccessibilityRole, IntHash<AccessibilityRole>, WTF::StrongEnumHashTraits<AccessibilityRole>>;
+
+ALWAYS_INLINE String accessibilityRoleToString(AccessibilityRole role)
```
this sort of looks like a lot of code to always inline. I'd have to imagine we use this method a lot. It's got to add some overheard to code bloat that's real I think

minorninth replied:
Would be nice to follow up by adding AccessibilityRoleInlines.h and then including those two files directly from only the files that actually need them.

twilco replied:
Removed the ALWAYS_INLINE, agreed that inlining this function seems wrong with how large it is.

rr-codes commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -1864,6 +1864,7 @@
 		48B187762E54CC85009B57A9 /* AXUtilities.h in Headers */ = {isa = PBXBuildFile; fileRef = 80A24B442E486831006FF9DB /* AXUtilities.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		48B187772E54CC8D009B57A9 /* AXObjectRareData.h in Headers */ = {isa = PBXBuildFile; fileRef = 48B187722E54CC65009B57A9 /* AXObjectRareData.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		48B187782E54CC95009B57A9 /* AXTableHelpers.h in Headers */ = {isa = PBXBuildFile; fileRef = 48B187742E54CC65009B57A9 /* AXTableHelpers.h */; settings = {ATTRIBUTES = (Private, ); }; };
+		48B187902E5CAE0F009B57A9 /* AccessibilityRole.h in Headers */ = {isa = PBXBuildFile; fileRef = 48B1878F2E5CAE0F009B57A9 /* AccessibilityRole.h */; settings = {ATTRIBUTES = (Private, ); }; };
```
Does this need to be private instead of project?

twilco replied:
Hmm, I followed the pattern I saw in other headers. Is there a reason to go with project over private?

rr-codes replied:
Yes; all header should always prefer to be project-level whenever possible. (in fact, it's even the default when adding a file to a target in Xcode!) This is to avoid exposing the header to private clients of WebCore (aka WebKit) when it's not needed to facilitate better encapsulation (and, as a side effect, also reduces the scope of the necessary module). 

The times where it's not possible for a header to be project is when it's used by WebKit, or when it's used by another header in WebCore that is itself a private header.

twilco replied:
That's helpful to know! I think this must be private then — `AccessibilityRole` is necessarily included by `AXCoreObject.h`, which is necessarily included by `AccessibilityObject.h`, which has usage outside WebCore.

If we want to try to untangle this, we should try in a follow-up I think.

rr-codes replied:
ah, that makes sense then!

rr-codes commented on Source/WebCore/accessibility/AccessibilityRole.h:
```diff
@@ -0,0 +1,167 @@
+
+namespace WebCore {
+
+enum class AccessibilityRole : uint8_t {
```
since this is using `uint8_t`, would you mind including `<cstdint>` in the header? (This will be needed to be done eventually for modularization, so it's easier if done now, but no worries if not! I'll clean it up eventually otherwise)

twilco replied:
Done!

twilco committed:
AX: Move AccessibilityRole into its own header to fix a non-unified build issue in AXTableHelpers.cpp
https://bugs.webkit.org/show_bug.cgi?id=297859
rdar://159107755

Reviewed by Joshua Hoffman.

This commit allows use of AccessibilityRole without importing the entirety of AXCoreObject.h.

Renames accessibilityRoleToString to roleToString.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::accessibilityRoleToString): Deleted.
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXTableHelpers.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::debugDescription const):
* Source/WebCore/accessibility/AXUtilities.cpp:
(WebCore::roleToString):
* Source/WebCore/accessibility/AXUtilities.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::debugDescriptionInternal const):
* Source/WebCore/accessibility/AccessibilityRole.h: Added.
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper _accessibilityWebRoleAsString]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::debugDescriptionInternal const):

Canonical link: https://commits.webkit.org/299307@main
